rclpy/rclpy/callback_groups.py

```diff
@@ -0,0 +1,107 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from threading import Lock
+import weakref
+
+
+class CallbackGroup:
+    """Control when callbacks are allowed to be executed."""
+
+    def __init__(self):
+        super().__init__()
+        self.entities = set()
+
+    def add_entity(self, entity):
+        """
+        Add an entity to the callback group.
+
+        entity - a subscription, timer, client, or service instance
+        """
+        self.entities.add(weakref.ref(entity))
+
+    def has_entity(self, entity):
+        """
+        Determine if an entity has been added to this group.
+
+        entity - a subscription, timer, client, or service instance
+        """
+        return weakref.ref(entity) in self.entities
+
+    def can_take(self, entity):
+        """
+        Return true if an entity can be executed.
+
+        entity - a subscription, timer, client, or service instance
+        """
+        raise NotImplementedError("Derived classes must implement this")
```

Comment:
I would say the exception type already describes the reason / intention. I would skip the custom message - but that is probably pretty subjective.

Anyway the string should be in single quotes.

Reply:
Message removed in fa60f5ab0be6b350a7e63b4e696bf7d5aca854e5